File: Robotic Controllers/src/main/Population.java
package main;

import java.util.Arrays;
import java.util.Collections;

public class Population {
    private Entity[] population;
    private double populationFitness = -1;

    public Population(int populationSize) {
        this.population = new Entity[populationSize];
    }

    public Population(int populationSize, int chromosomeLenght) {
        this(populationSize);

        for (int i = 0; i < populationSize; i++) {
            Entity entity = new Entity(chromosomeLenght);
            this.population[i] = entity;
        }
    }

    public Entity[] getEntities() {
        return this.population;
    }

    public Entity getFittest(int index) {
        Arrays.sort(this.population, (o1, o2) -> Double.compare(o2.getFitness(), o1.getFitness()));
        return population[index];
    }

    public double getPopulationFitness() {
        return populationFitness;
    }

    public void setPopulationFitness(double populationFitness) {
        this.populationFitness = populationFitness;
    }

    public int size() {
        return this.population.length;
    }

    public Entity setEntity(int idx, Entity entity) {
        return this.population[idx] = entity;
    }

    public Entity getEntity(int idx) {
        return this.population[idx];
    }

    public void shuffle() {
        Collections.shuffle(Arrays.asList(population));
    }
}

File: Genetic K-means/models/population.py
from sklearn import datasets

from entity import Entity
import numpy as np


class Population:
	def __init__(self, n_clusters: int, population_size: int, X: np.array):
		"""

		:param n_clusters: The number of clusters
		:param population_size: The number of individuals in the population
		:param X: The data we are trying to cluster
		"""
		self.population_size = population_size
		self.population = []
		self.n_clusters = n_clusters
		self.population_fitness = -1
		self.init_population(X)

	def init_population(self, X: np.array):
		for i in range(self.population_size):
			self.population.append(Entity(self.n_clusters, X))

	def get_entities(self):
		return self.population

	def get_size(self):
		return self.population_size

	def set_fitness(self, fitness):
		self.population_fitness = fitness

	def get_pop_fitness(self):
		return self.population_fitness

	def set_entity(self, index: int, entity: Entity):
		self.population[index] = entity

	def get_entity(self, index: int):
		return self.population[index]

	def get_fittest(self):
		self.population.sort(key=lambda entity: entity.get_fitness())
		return self.population[0]

	def __str__(self):
		output = ""
		for entity in self.population:
			output = output + entity.to_string() + "\n"
		return output

	def to_string(self):
		return self.__str__()


def main():
	iris = datasets.load_iris()
	X = iris.data
	y = iris.target

	pop = Population(3, 5, X)

	print("decoded", pop.get_entity(0).decode())

	print(pop)


if __name__ == '__main__':
	main()
